### PR TITLE
[FLINK-7209] [table] Support DataView in Tuple and Case Class as the ACC type of AggregateFunction

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request adds support of DataView in Tuple and Case Class as the accumulator type of AggregateFunction*

## Brief change log

  - *Updates UserDefinedFunctionUtils#removeStateViewFieldsFromAccTypeInfo to process TupleTypeInfo and CaseClassTypeInfo for DataView*
  - *Updates MapViewTypeInfo and ListViewTypeInfo to use NullSerializer for serialization when state backend is used*

## Verifying this change

This change added tests and can be verified as follows:

  - *Updated tests of AggregateITCase#testGroupAggregateWithStateBackend*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.dataview
+
+import org.apache.flink.api.common.typeutils.base.TypeSerializerSingleton
+import org.apache.flink.core.memory.{DataInputView, DataOutputView}
+
+class NullSerializer extends TypeSerializerSingleton[Any] {
```
Comment (walterddr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala]: I am assuming `NullSerializer` is needed because the actual map is actually handled by StateBackend and because we removed them through `removeStateViewFieldsFromAccTypeInfo`. I was wondering why is this needed since it was not introduced in previous approach. Neither was it reflected in the test case ( still passes without the nullserializer setting)

Comment (dianfu) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala]: I think the reason `NullSerializer` is not needed previously is the magic of PojoTypeInfo(PojoTypeSerializer). PojoTypeInfo requires to specify the PojoFields for all the fields we want to serialize. For a DataView field, when we remove it through `removeStateViewFieldsFromAccTypeInfo`, we have also constructed a new PojoTypeInfo which removed the corresponding `PojoField` of the DataView field. While for TupleTypeInfo and CaseClassTypeInfo, they have no such magic.

**Comment by dianfu** [comment]
@walterddr Thanks a lot for your review. Agree with you that all the code paths should be well tested. This PR only adds some test cases for CaseClass/Tuple as the accumulator type as other code paths can be well tested by existing test cases. Please let me know if you found any code paths which are not tested. Thanks again!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 2]
```diff
@@ -489,6 +491,88 @@ object UserDefinedFunctionUtils {
...
+            "agg" + index + "$" + fieldName,
+            field,
+            newTypeInfo)
+
+          (newTypeInfo, Some(spec))
+        } else {
+          (newTypeInfo, None)
+        }
+
+      case list: ListViewTypeInfo[_] =>
```
[the remark this answers is not in the document]
Comment (walterddr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: Sorry I was not making it clear. I meant maybe we can at a `Tuple<ListView, Long>` test case.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala (hunk 1 above)
Comment (walterddr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala]: Yeah. I think I also trace the same code path for Pojo. However, removing the `NullSerializer` still lets the newly modified `testGroupAggregateWithStateBackend` passes. this is worrying because essentially there's no test safeguard for this piece of code being removed in the future.

Comment (dianfu) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala]: The tests can still pass without NullSerializer is because NullSerializer is an optimization. It doesn't affect the correctness. So an IT case can not cover this kind of tests. May be we should add a unit test for UserDefinedFunctionUtil#removeStateViewFieldsFromAccTypeInfo. But I found that currently there is no unit test for UserDefinedFunctionUtil. What do you think about adding a unit test class for UserDefinedFunctionUtil?

Comment (walterddr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala]: hmm. I think that's a good idea but should probably be addressed in another JIRA. I actually have [PR:6472](https://github.com/apache/flink/pull/6472) still pending which I am also very fond of the idea to add test for the Utility class.

Comment (dianfu) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/NullSerializer.scala]: What about only adding UserDefinedFunctionUtilTest#testRemoveStateViewFieldsFromAccTypeInfo in this PR? Just like you said, the tests of other methods of UserDefinedFunctionUtil can be done in another JIRA if needed.
